### ISSUE TITLE
Issue with creating a custom module for data loading

### ISSUE THREAD

**Issue by s-cork** [olifolkerd/tabulator#3621]
**Describe the bug**
When a table is initiated, 
`this._clearObjectPointers();` is called after setting the options.
The problem is that the data is assumed to be either a string or an array - `this.options.data = this.options.data.slice(0);`
But Tabulator 5 allows me to provide my own data loaders - so long as I return a Promise that resolves to an Array of objects in the `data-load` event.
i.e. I should be able to initiate the data option however I want so long as the `data-load` chained return value conforms.

**Tabulator Info**
- 5.1.2

**Working Example**
https://codepen.io/s-cork/pen/abVaWRo

**To Reproduce**
A step by step guide to recreate the issue in your JS Fiddle or Codepen:
1. In the example above open the browser console
2. observe the error message
```
Uncaught TypeError: this.options.data.slice is not a function
    at Tabulator._clearObjectPointers (tabulator-tables.js:16932:45)
    at Tabulator.initialzeCoreSystems (tabulator-tables.js:16879:10)
    at new Tabulator (tabulator-tables.js:16848:12)
    at index.html:60:16
```

**Expected behavior**
The `_clearObjectPointers` should check if data is an Array or not.
But even then the custom module might rely on data being the original data like reactiveData does.

**Comment by olifolkerd** [comment]
Hey @s-cork 

In this case i would still consider a function to be an inappropriate value for the data loader. For your usage case if you are building a custom module like this i would suggest that you register your own option like `dataLoadFunc` or something that you can then pass your data too.  you can do this using the `registerTableOption` function in the module constructor. for full details see the [Building Modules Documentation](http://tabulator.info/docs/5.1/module-build#options)

The ajax module works in a similar why where is uses the `ajaxUrl` option for the URL not the `data` option itself.

You will see a similar setup in the [Database Module Example](http://tabulator.info/docs/5.1/module-examples#db)

Cheers

Oli :)

**Comment by s-cork** [comment]
Yeah that make sense. 
I guess I got confused because with Ajax you do `setData(url)`.
So in my example I'd still be able to do `setData(func)`, which is ok?

i.e. really it's 
> I can make `setData()` behave however I want so long as the return value from `data-load` conforms.

**Comment by olifolkerd** [comment]
Hey @s-cork 

You are adding your own module, you can add any functions you like to the table with that module. you should change the paradigm you are thinking in there. 

If you need it to do something different from the normal operation of the table, then you should have a different function. it can then call the core table functionality from the module when it has the info it needs.

Cheers

Oli :)

**Comment by s-cork** [comment]
Hi @olifolkerd,

I just wanted to have another crack at this issue, because I was looking through more of the source code.
And the `ImportModule` supports exactly the type of pattern I'm proposing here.

I basically want to write my own `ImportModule`.
I get the paradigm shift you're talking about, i.e that I can do my own tableOption and tableFunction, and ask my uses to use those,
but given that the importModule uses the paradigm i'm after, I think it's an argument in favour of this issue 🙃

i.e. the `ImportModule` says it's fine to set the `data` option to a string and then I'll handle it.
Which means in the `_clearObjectPointers` there's this call to `.slice()` on a string which is unnecessary given that it's a string.

**Comment by olifolkerd** [comment]
Hey @s-cork 

At the point you are going to write your own import module, why not just write your own importer that can use that module instead?

While i would consider allowing other types of data to be passed into the data for the importer as i do see a point there, and would open up the values it can take, but i would certainly not be looking to make it call functions if they are set.  I still feel that what you are proposing is a total anti-pattern for that prop.

The `data` props purpose is to take the initial data for the table, not to calculate it. it is taking purely static data, and you are attempting to overload it to take dynamic data. If you dont have the data straight away, you should fetch it before instantiating the table. This functionality will not be changing

As I pointed out before, with modules that take async data they have their own prop that is designed to handle this change

Cheers

Oli

**Comment by s-cork** [comment]
Ah - my use case isn't actually a function - that was just to demonstrate the issue.

My users have complex javascript data structures since I write a wrapper for Tabulator within a python compiled to javascript environment!
Very niche!
So the objects I want to pass as data are python javascript wrapped object that I would then cleanup.
i.e. they can't be sliced 😅

(That's why I reported #3415)

**Comment by olifolkerd** [comment]
Hey @s-cork 

Ahh, ok, now that makes a great deal more sense.

Ok no probs, i will add this as a feature as it will involve a change to existing functionality that i would like to document in more detail.

Cheers

Oli :)

**Comment by olifolkerd** [comment]
This was included in the 5.1.7 patch release.

Cheers

Oli :)
